examples/vision/knowledgedistillation.py

```diff
@@ -0,0 +1,224 @@
+"""
+Title: Knowledge Distillation
+Author: [Kenneth Borup](https://twitter.com/Kennethborup)
+Date created: 2020/09/01
+Last modified: 2020/09/01
+Description: Implementation of classical Knowledge Distillation.
+"""
+
+"""
+## Introduction to Knowledge Distillation
+
+Knowledge Distillation (or student-teacher training) is a procedure for model
+compression, in which a small (student) model is trained to match a large pre-trained
+(teacher) model. Intrinsic knowledge is transferred from the teacher model to the student
+by minimizing a loss function, aimed at matching softened teacher logits as well as
+ground-truth labels.
+The logits are softened by applying a temperature in the softmax, effectively smoothing
+out the probability distribution and revealing inter-class relationships learned by the
+teacher.
+
+- [Paper](https://arxiv.org/abs/1503.02531)
+"""
+
+"""
+## Setup
+"""
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+import numpy as np
+
+"""
+## Construct `Distiller()` class
+The custom `Distiller()` class, merges custom `train_step` and `test_step` with classical
+and custom Keras `compile()` elements. In order to apply the distiller, we need:
+
+- A trained teacher model
+- A student model to train
+- A student loss function on the difference between student predictions and ground-truth
+- A distillation loss function, along with a temperature `T`, on the difference between
+the soft student predictions and the soft teacher labels
+- `alpha` to weight the student and distillation loss
+- An optimizer for the student and (optional) metrics to evaluate performance
+
+In `train_step` we perform a forward pass of both the teacher and student, calculate the
+loss with weighting of the student loss `s_loss` and distillation loss `d_loss` by
+`alpha` and `1-alpha`, respectively, and update weights. Note, only the student weights
+are updated and therefore, we merely calculate the gradients for the student weights. In
+the `test_step` we evaluate the student model on the provided dataset as usual.
+"""
+
+
+class Distiller(keras.Model):
+    def __init__(self, student, teacher):
+        super(Distiller, self).__init__()
+        self.teacher = student
+        self.student = teacher
+
+    def compile(self, optimizer, metrics, s_loss_fn, d_loss_fn, alpha, T):
```

Comment:
Please add a docstring with argument descriptions.

Prefer fully-spelled out lower case snake case variable names, e.g. `temperature` and `student_loss`.

`alpha` and `temperature` are arguments that should have a default, e.g. 0.1 and 3.